src/DataTypes/convertMySQLDataType.cpp

```diff
@@ -25,48 +28,69 @@ ASTPtr dataTypeConvertToQuery(const DataTypePtr & data_type)
     return makeASTFunction("Nullable", dataTypeConvertToQuery(typeid_cast<const DataTypeNullable *>(data_type.get())->getNestedType()));
 }
 
-DataTypePtr convertMySQLDataType(const std::string & mysql_data_type, bool is_nullable, bool is_unsigned, size_t length)
+DataTypePtr convertMySQLDataType(const std::string & mysql_data_type, bool is_nullable, bool is_unsigned, size_t length, size_t precision, size_t scale)
 {
+    // we expect mysql_data_type to be either "basic_type" or "type_with_params(param1, param2, ...)"
+    auto data_type = std::string_view(mysql_data_type);
+    const auto param_start_pos = data_type.find("(");
+    const auto type_name = data_type.substr(0, param_start_pos);
+
     DataTypePtr res;
-    if (mysql_data_type == "tinyint")
+    if (type_name == "tinyint")
     {
         if (is_unsigned)
             res = std::make_shared<DataTypeUInt8>();
         else
             res = std::make_shared<DataTypeInt8>();
     }
-    else if (mysql_data_type == "smallint")
+    else if (type_name == "smallint")
     {
         if (is_unsigned)
             res = std::make_shared<DataTypeUInt16>();
         else
             res = std::make_shared<DataTypeInt16>();
     }
-    else if (mysql_data_type == "int" || mysql_data_type == "mediumint")
+    else if (type_name == "int" || type_name == "mediumint")
     {
         if (is_unsigned)
             res = std::make_shared<DataTypeUInt32>();
         else
             res = std::make_shared<DataTypeInt32>();
     }
-    else if (mysql_data_type == "bigint")
+    else if (type_name == "bigint")
     {
         if (is_unsigned)
             res = std::make_shared<DataTypeUInt64>();
         else
             res = std::make_shared<DataTypeInt64>();
     }
-    else if (mysql_data_type == "float")
+    else if (type_name == "float")
         res = std::make_shared<DataTypeFloat32>();
-    else if (mysql_data_type == "double")
+    else if (type_name == "double")
         res = std::make_shared<DataTypeFloat64>();
-    else if (mysql_data_type == "date")
+    else if (type_name == "date")
         res = std::make_shared<DataTypeDate>();
-    else if (mysql_data_type == "datetime" || mysql_data_type == "timestamp")
+    if (type_name == "timestamp" && scale == 0)
+    {
         res = std::make_shared<DataTypeDateTime>();
-    else if (mysql_data_type == "binary")
+    }
+    else if (type_name == "datetime" || type_name == "timestamp")
```

Comment:
does it mean that simple old-good datetime will be converted now to DateTime64?

In mysql versions <= 5.5 (afaik) it doesn't support scale parameter at all. 
http://sqlfiddle.com/#!9/53e7cec/1

Reply:
Maybe it's worth to test with some old MySQL version (i think manual test should be enough - just try to change mysql docker version)

Reply:
Reason for choosing DateTime64 is the range of MySQL's `DATETIME`, which doesn't fit into `DateTime` underlying storage type:

> The DATETIME type is used for values that contain both date and time parts. MySQL retrieves and displays DATETIME values in 'YYYY-MM-DD hh:mm:ss' format. The supported range is '1000-01-01 00:00:00' to '9999-12-31 23:59:59'. 

https://docs.oracle.com/cd/E17952_01/mysql-5.5-en/datetime.html
https://dev.mysql.com/doc/refman/5.7/en/datetime.html

Reply:
It looks like backward incompatible change. Should it return the same types in results in partially updated CH cluster? We need a setting disabled by default if not.

Reply:
Added a setting `mysql_datatypes_support_level` which can be empty or any combination of `decimal` or `datetime64`, (`enum` is coming later), e.g.:
* `'decimal,datetime64'`
* `'datetime64'`
* `'decimal'`
* `''`